<a href="https://colab.research.google.com/github/lovishkanther11/Deep-Learning/blob/master/Keras_losses.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Usage of loss functions

A loss function (or objective function, or optimization score function) is one of the two parameters required to compile a model:

```
model.compile(loss='mean_squared_error', optimizer='sgd')

from keras import losses
model.compile(loss=losses.mean_squared_error, optimizer='sgd')
```

You can either pass the name of an existing loss function, or pass a TensorFlow/Theano symbolic function that returns a scalar for each data-point and takes the following two arguments:

1. y_true: True labels. TensorFlow/Theano tensor.
2. y_pred: Predictions. TensorFlow/Theano tensor of the same shape as y_true.

The actual optimized objective is the mean of the output array across all datapoints.

In [1]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], NUM_ROWS * NUM_COLS))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], NUM_ROWS * NUM_COLS))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)


# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='elu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='elu'))
model.add(Dense(10, activation='softmax'))



Using TensorFlow backend.


11493376/11490434 [==============================] - 0s 0us/step





#mean_squared_error

keras.losses.mean_squared_error(y_true, y_pred)

In [2]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='mean_squared_error',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])





Train on 60000 samples, validate on 10000 samples
Epoch 1/10





60000/60000 [==============================] - 11s 189us/step - loss: 0.0163 - acc: 0.8897 - val_loss: 0.0102 - val_acc: 0.9352
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0073 - acc: 0.9534 - val_loss: 0.0071 - val_acc: 0.9532
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0053 - acc: 0.9659 - val_loss: 0.0054 - val_acc: 0.9657
Epoch 4/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0042 - acc: 0.9733 - val_loss: 0.0049 - val_acc: 0.9697
Epoch 5/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0036 - acc: 0.9778 - val_loss: 0.0042 - val_acc: 0.9751
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0031 - acc: 0.9803 - val_loss: 0.0042 - val_acc: 0.9746
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0027 - acc: 0.9829 - val_loss: 0.003

#mean_absolute_error

keras.losses.mean_absolute_error(y_true, y_pred)

In [3]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='mean_absolute_error',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0025 - acc: 0.9892 - val_loss: 0.0042 - val_acc: 0.9808
Epoch 2/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0022 - acc: 0.9901 - val_loss: 0.0048 - val_acc: 0.9773
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.0020 - acc: 0.9912 - val_loss: 0.0041 - val_acc: 0.9803
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0018 - acc: 0.9917 - val_loss: 0.0045 - val_acc: 0.9782
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0016 - acc: 0.9926 - val_loss: 0.0041 - val_acc: 0.9808
Epoch 6/10
60000/60000 [==============================] - 2s 26us/step - loss: 0.0016 - acc: 0.9929 - val_loss: 0.0041 - val_acc: 0.9802
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0014 - acc: 0.9934 - val_loss: 0.0040 - val_acc

#mean_absolute_percentage_error

keras.losses.mean_absolute_percentage_error(y_true, y_pred)

In [4]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='mean_absolute_percentage_error',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 35us/step - loss: 612911.9891 - acc: 0.9942 - val_loss: 1996683.4985 - val_acc: 0.9809
Epoch 2/10
60000/60000 [==============================] - 2s 34us/step - loss: 594180.5650 - acc: 0.9943 - val_loss: 1852941.2381 - val_acc: 0.9820
Epoch 3/10
60000/60000 [==============================] - 2s 32us/step - loss: 552259.4252 - acc: 0.9947 - val_loss: 1712752.9777 - val_acc: 0.9836
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: 529650.0710 - acc: 0.9950 - val_loss: 1894412.1706 - val_acc: 0.9818
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 558876.7179 - acc: 0.9947 - val_loss: 1856284.4753 - val_acc: 0.9817
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 499283.8617 - acc: 0.9951 - val_loss: 1876335.2922 - val_acc: 0.9819
Epoch 7/10
60000/60000 [==============================] - 2s 2

#mean_squared_logarithmic_error

keras.losses.mean_squared_logarithmic_error(y_true, y_pred)

In [5]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='mean_squared_logarithmic_error',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])


Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 36us/step - loss: 4.1255e-04 - acc: 0.9954 - val_loss: 0.0015 - val_acc: 0.9825
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 3.8047e-04 - acc: 0.9958 - val_loss: 0.0017 - val_acc: 0.9805
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 4.0731e-04 - acc: 0.9954 - val_loss: 0.0018 - val_acc: 0.9795
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 3.9885e-04 - acc: 0.9955 - val_loss: 0.0017 - val_acc: 0.9815
Epoch 5/10
60000/60000 [==============================] - 2s 28us/step - loss: 4.0987e-04 - acc: 0.9954 - val_loss: 0.0018 - val_acc: 0.9793
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 3.7982e-04 - acc: 0.9957 - val_loss: 0.0016 - val_acc: 0.9815
Epoch 7/1

#squared_hinge

keras.losses.squared_hinge(y_true, y_pred)

In [6]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='squared_hinge',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.9004 - acc: 0.9956 - val_loss: 0.9017 - val_acc: 0.9815
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.9004 - acc: 0.9960 - val_loss: 0.9018 - val_acc: 0.9809
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.9004 - acc: 0.9960 - val_loss: 0.9019 - val_acc: 0.9793
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.9004 - acc: 0.9962 - val_loss: 0.9017 - val_acc: 0.9827
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.9004 - acc: 0.9962 - val_loss: 0.9018 - val_acc: 0.9815
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.9003 - acc: 0.9964 - val_loss: 0.9018 - val_acc: 0.9814
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.9004 - acc: 0.9962 - val_loss: 0.9018 - val_acc

#hinge

keras.losses.hinge(y_true, y_pred)

In [7]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='hinge',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.9003 - acc: 0.9968 - val_loss: 0.9017 - val_acc: 0.9829
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.9003 - acc: 0.9968 - val_loss: 0.9017 - val_acc: 0.9825
Epoch 3/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.9003 - acc: 0.9970 - val_loss: 0.9018 - val_acc: 0.9820
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.9003 - acc: 0.9970 - val_loss: 0.9020 - val_acc: 0.9803
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.9003 - acc: 0.9971 - val_loss: 0.9018 - val_acc: 0.9822
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.9003 - acc: 0.9971 - val_loss: 0.9017 - val_acc: 0.9830
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.9003 - acc: 0.9971 - val_loss: 0.9016 - val_acc

#categorical_hinge

keras.losses.categorical_hinge(y_true, y_pred)

In [8]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='categorical_hinge',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 38us/step - loss: 0.0056 - acc: 0.9972 - val_loss: 0.0387 - val_acc: 0.9806
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0057 - acc: 0.9972 - val_loss: 0.0342 - val_acc: 0.9831
Epoch 3/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0052 - acc: 0.9974 - val_loss: 0.0348 - val_acc: 0.9825
Epoch 4/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0057 - acc: 0.9971 - val_loss: 0.0330 - val_acc: 0.9836
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0057 - acc: 0.9972 - val_loss: 0.0336 - val_acc: 0.9831
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0051 - acc: 0.9975 - val_loss: 0.0359 - val_acc: 0.9824
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0054 - acc: 0.9973 - val_loss: 0.0363 - val_acc

#logcosh

keras.losses.logcosh(y_true, y_pred)

Logarithm of the hyperbolic cosine of the prediction error.

log(cosh(x)) is approximately equal to (x ** 2) / 2 for small x and to abs(x) - log(2) for large x. This means that 'logcosh' works mostly like the mean squared error, but will not be so strongly affected by the occasional wildly incorrect prediction.

##Arguments

1. y_true: tensor of true targets.
2. y_pred: tensor of predicted targets.

##Returns

1. Tensor with one scalar loss entry per sample.

In [9]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='logcosh',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 37us/step - loss: 2.3037e-04 - acc: 0.9972 - val_loss: 0.0015 - val_acc: 0.9827
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 2.3597e-04 - acc: 0.9971 - val_loss: 0.0015 - val_acc: 0.9824
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 2.3819e-04 - acc: 0.9971 - val_loss: 0.0015 - val_acc: 0.9817
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 2.2481e-04 - acc: 0.9973 - val_loss: 0.0016 - val_acc: 0.9806
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 2.2506e-04 - acc: 0.9973 - val_loss: 0.0015 - val_acc: 0.9819
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 2.3015e-04 - acc: 0.9972 - val_loss: 0.0016 - val_acc: 0.9809
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 2.3075e-04 - acc: 0.9972 

#categorical_crossentropy

keras.losses.categorical_crossentropy(y_true, y_pred, from_logits=False, label_smoothing=0)

In [10]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 38us/step - loss: 0.0323 - acc: 0.9975 - val_loss: 0.1832 - val_acc: 0.9836
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0341 - acc: 0.9974 - val_loss: 0.1931 - val_acc: 0.9831
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0320 - acc: 0.9977 - val_loss: 0.1893 - val_acc: 0.9835
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0314 - acc: 0.9978 - val_loss: 0.1989 - val_acc: 0.9819
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0314 - acc: 0.9977 - val_loss: 0.2100 - val_acc: 0.9816
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0311 - acc: 0.9977 - val_loss: 0.1958 - val_acc: 0.9817
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0307 - acc: 0.9978 - val_loss: 0.1972 - val_acc

#binary_crossentropy

keras.losses.binary_crossentropy(y_true, y_pred, from_logits=False, label_smoothing=0)

In [11]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 3s 42us/step - loss: 0.0060 - acc: 0.9995 - val_loss: 0.0418 - val_acc: 0.9960
Epoch 2/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0062 - acc: 0.9995 - val_loss: 0.0399 - val_acc: 0.9964
Epoch 3/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0065 - acc: 0.9995 - val_loss: 0.0395 - val_acc: 0.9965
Epoch 4/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0064 - acc: 0.9995 - val_loss: 0.0468 - val_acc: 0.9958
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0062 - acc: 0.9995 - val_loss: 0.0434 - val_acc: 0.9960
Epoch 6/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0062 - acc: 0.9995 - val_loss: 0.0371 - val_acc: 0.9963
Epoch 7/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0057 - acc: 0.9996 - val_loss: 0.0367 - val_acc

#kullback_leibler_divergence

keras.losses.kullback_leibler_divergence(y_true, y_pred)

In [12]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='kullback_leibler_divergence',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 38us/step - loss: 0.0293 - acc: 0.9979 - val_loss: 0.1939 - val_acc: 0.9827
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0272 - acc: 0.9981 - val_loss: 0.2177 - val_acc: 0.9812
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0267 - acc: 0.9981 - val_loss: 0.2002 - val_acc: 0.9837
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0281 - acc: 0.9980 - val_loss: 0.2117 - val_acc: 0.9820
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0273 - acc: 0.9980 - val_loss: 0.2200 - val_acc: 0.9810
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0265 - acc: 0.9981 - val_loss: 0.2002 - val_acc: 0.9829
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0256 - acc: 0.9983 - val_loss: 0.1949 - val_acc

#poisson

keras.losses.poisson(y_true, y_pred)

In [13]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='poisson',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 39us/step - loss: 0.1027 - acc: 0.9980 - val_loss: 0.1200 - val_acc: 0.9824
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1025 - acc: 0.9983 - val_loss: 0.1198 - val_acc: 0.9840
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.1025 - acc: 0.9983 - val_loss: 0.1202 - val_acc: 0.9821
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.1026 - acc: 0.9982 - val_loss: 0.1215 - val_acc: 0.9825
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.1027 - acc: 0.9981 - val_loss: 0.1212 - val_acc: 0.9824
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.1026 - acc: 0.9983 - val_loss: 0.1200 - val_acc: 0.9830
Epoch 7/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1026 - acc: 0.9982 - val_loss: 0.1200 - val_acc

#cosine_proximity

keras.losses.cosine_proximity(y_true, y_pred, axis=-1)

In [14]:
# Compile model using accuracy
model.compile(optimizer='rmsprop',
              loss='cosine_proximity',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 41us/step - loss: -0.9982 - acc: 0.9981 - val_loss: -0.9827 - val_acc: 0.9823
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: -0.9984 - acc: 0.9984 - val_loss: -0.9823 - val_acc: 0.9819
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: -0.9981 - acc: 0.9981 - val_loss: -0.9818 - val_acc: 0.9815
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: -0.9982 - acc: 0.9981 - val_loss: -0.9822 - val_acc: 0.9820
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: -0.9981 - acc: 0.9980 - val_loss: -0.9814 - val_acc: 0.9809
Epoch 6/10
60000/60000 [==============================] - 2s 30us/step - loss: -0.9982 - acc: 0.9981 - val_loss: -0.9813 - val_acc: 0.9809
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: -0.9982 - acc: 0.9982 - val_loss: -0.

#Thank you for completing this notebook

1)**mean_squared_error**

Test loss: 0.003352416222123199


Test accuracy: 0.9799

2)**mean_absolute_error**

Test loss: 0.0038889680898445095


Test accuracy: 0.981

3)**mean_absolute_percentage_error**

Test loss: 1803200.477230715


Test accuracy: 0.9819

4)**mean_squared_logarithmic_error**

Test loss: 0.0017203592830507266


Test accuracy: 0.9807

5)**squared_hinge**

Test loss: 0.9018609205245972


Test accuracy: 0.9803

6)**hinge**

Test loss: 0.901724101448059


Test accuracy: 0.9833

7)**categorical_hinge**

Test loss: 0.03607132461245253


Test accuracy: 0.982

8)**logcosh**

Test loss: 0.0015778261381334592


Test accuracy: 0.981

9)**categorical_crossentropy**

Test loss: 0.23053656777290435


Test accuracy: 0.9802

10)**binary_crossentropy**

Test loss: 0.04225646639997729


Test accuracy: 0.996229995727539

11)**kullback_leibler_divergence**

Test loss: 0.21674129924554209


Test accuracy: 0.9822

12)**poisson**

Test loss: 0.1220341134428978


Test accuracy: 0.9826

13)**cosine_proximity**

Test loss: -0.9826041687011718


Test accuracy: 0.9819

The best among all is **binary_crossentropy** with an accuracy of almost 1.

In [15]:
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, Y_train), (X_test, Y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(Y_train, 10)
Y_test =  to_categorical(Y_test, 10)

#Model building
model = Sequential()
model.add(Dense(1024, input_shape=(3072, )))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(10))
model.add(Activation('softmax'))

# Compile model using above optimizer
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Exception: ignored

**CIFAR 10**

Test loss: 0.23143027470111846


Test accuracy: 0.9121199871063232

In [16]:
from keras.datasets import cifar100
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, Y_train), (X_test, Y_test) = cifar100.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(Y_train, 100)
Y_test =  to_categorical(Y_test, 100)

#Model building
model = Sequential()
model.add(Dense(1024, input_shape=(3072, )))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(100))
model.add(Activation('softmax'))


from keras import optimizers
adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)

# Compile model using above optimizer
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

169009152/169001437 [==============================] - 3s 0us/step
Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 4s 75us/step - loss: 0.0584 - acc: 0.9897 - val_loss: 0.0485 - val_acc: 0.9900
Epoch 2/10
50000/50000 [==============================] - 3s 57us/step - loss: 0.0478 - acc: 0.9900 - val_loss: 0.0464 - val_acc: 0.9900
Epoch 3/10
50000/50000 [==============================] - 3s 59us/step - loss: 0.0456 - acc: 0.9900 - val_loss: 0.0458 - val_acc: 0.9900
Epoch 4/10
50000/50000 [==============================] - 3s 57us/step - loss: 0.0442 - acc: 0.9901 - val_loss: 0.0457 - val_acc: 0.9900
Epoch 5/10
50000/50000 [==============================] - 3s 57us/step - loss: 0.0430 - acc: 0.9902 - val_loss: 0.0499 - val_acc: 0.9897
Epoch 6/10
50000/50000 [==============================] - 3s 58us/step - loss: 0.0421 - acc: 0.9902 - val_loss: 0.0474 - val_acc: 0.9898
Epoch 7/10
50000/50000 [==============================] - 3s 

**CIFAR 100**

Test loss: 0.04386511961221695


Test accuracy: 0.990127011680603

In [17]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import fashion_mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, Y_train), (X_test, Y_test) = fashion_mnist.load_data()



# Reshape data
X_train = X_train.reshape((X_train.shape[0], NUM_ROWS * NUM_COLS))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], NUM_ROWS * NUM_COLS))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
Y_train = to_categorical(Y_train, NUM_CLASSES)
Y_test = to_categorical(Y_test, NUM_CLASSES)


# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='relu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

# Compile model using above optimizer
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

4423680/4422102 [==============================] - 1s 0us/step
Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 41us/step - loss: 0.0943 - acc: 0.9622 - val_loss: 0.0933 - val_acc: 0.9592
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0672 - acc: 0.9728 - val_loss: 0.0739 - val_acc: 0.9703
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0603 - acc: 0.9759 - val_loss: 0.0622 - val_acc: 0.9762
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0561 - acc: 0.9775 - val_loss: 0.0649 - val_acc: 0.9746
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0528 - acc: 0.9787 - val_loss: 0.0682 - val_acc: 0.9749
Epoch 6/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0505 - acc: 0.9800 - val_loss: 0.0634 - val_acc: 0.9749
Epoch 7/10
60000/60000 [==============================] - 2s 31us

**FASHION_MNIST**

Test loss: 0.06934604954719543


Test accuracy: 0.9760300033569336